### PR TITLE
Refactor LoadMethod.LOCAL to use symlinks instead of copying directory

### LINKED ISSUE #614: Refactor LoadMethod.LOCAL to use symlinks as opposed to copying the dbt source dir
As of Cosmos 1.0-1.2, it copies the entire dbt project for every operator that runs dbt:
https://github.com/astronomer/astronomer-cosmos/blob/c82e6bc66dd31d6e1404fc24548d978489f4f0ab/cosmos/operators/local.py#L198

As reported in the #airflow-dbt slack, this can lead to performance issues:
https://apache-airflow.slack.com/archives/C059CC42E9W/p1697718075031609?thread_ts=1697708239.918849&cid=C059CC42E9W

We successfully adopted a strategy of using symbolic links from a temporary folder to run `dbt ls`:
https://github.com/astronomer/astronomer-cosmos/blob/c82e6bc66dd31d6e1404fc24548d978489f4f0ab/cosmos/dbt/graph.py#L167-L171

The goal with this ticket is to adopt this same strategy when running tasks using `LoadMethod.LOCAL`.

### PR DESCRIPTION
**Pull request by jbandoro**
## Description

This PR refactors the `create_symlinks` function that was previously used in load via dbt ls so that it can be used in `DbtLocalBaseOperator.run_command` instead of copying the entire directory.

## Related Issue(s)

closes #614 

## Breaking Change?

None

## Checklist

- [x] I have made corresponding changes to the documentation (if required)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by jbandoro** [comment]
It looks like symlinking this file: https://github.com/astronomer/astronomer-cosmos/blob/main/dev/dags/dbt/simple/data/imdb.db is locking the database causing the Sqlite integration test to fail. I think a workaround might be to ignore `.db` files.

File: cosmos/utils.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from pathlib import Path
```
Comment (tatiana) [inline cosmos/utils.py:1]: Let's avoid introducing generic modules at the root of Cosmos. I've seen `utils.py` becoming huge files with all sorts of things in previous projects.
What do you think if we add this to `dbt/project.py` instead? I'm open to other suggestions!

**Review by tatiana — COMMENTED** [review]
@jbandoro this is looking very good, one comment regarding the issue and only one comment regarding the code. Great job!

**Comment by tatiana** [comment]
We're really close to getting this merged, @jbandoro ! Please, let us know if we can support you in any sense

**Comment by jbandoro** [comment]
> We're really close to getting this merged, @jbandoro ! Please, let us know if we can support you in any sense

Thanks @tatiana! I've been struggling to get the sqlite integration test to pass with this refactor. I tried moving the `imdb.db` file outside of the simple dbt project, and updating the references [here where it sets the env var](https://github.com/astronomer/astronomer-cosmos/blob/6e41471712d9352524e70a59a1380e8e52c5ede2/dev/dags/example_cosmos_sources.py#L29C1-L29C1) and [here in the profiles.yml file](https://github.com/astronomer/astronomer-cosmos/blob/6e41471712d9352524e70a59a1380e8e52c5ede2/dev/dags/dbt/simple/profiles.yml#L10-L11) but they keep failing with `no such table: imdb.movies_ratings`.

I tried to move the sqlite db file on the main branch to see if it was related still to the symbolic linking and it results in the same error for me locally. 

If you have time to investigate I would be 🙏 otherwise I can look into it further.

File: cosmos/utils.py (hunk 1 above)
Comment (jbandoro) [reply cosmos/utils.py]: This has been updated, agree that a generic `utils` can be bloated - thanks for the suggestion.

**Comment by jbandoro** [comment]
Here's a the branch diff for the changes I tried to make: https://github.com/jbandoro/astronomer-cosmos/compare/use-symlinks-local-operator...jbandoro:astronomer-cosmos:test-moving-sqlite-file?expand=1 
and the error log from the sqlite integration test:
```shell
[2023-11-10T16:20:08.805-0800] {subprocess.py:73} INFO - (astronomer-cosmos) - Running command: ['/Users/justin.bandoro/Library/Application Support/hatch/env/virtual/astronomer-cosmos/3NzuKo9Z/tests.py3.10-2.6/bin/dbt', 'run', '--models', 'movies_ratings_simplified', '--profiles-dir', '/Users/justin.bandoro/astronomer-cosmos/dev/dags/dbt/simple', '--profile', 'simple', '--target', 'dev']
[2023-11-10T16:20:08.812-0800] {subprocess.py:84} INFO - (astronomer-cosmos) - Command output:
[2023-11-10T16:20:08.812-0800] {subprocess.py:84} INFO - (astronomer-cosmos) - Command output:
[2023-11-10T16:20:11.741-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:10  Running with dbt=1.4.0
[2023-11-10T16:20:11.741-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:10  Running with dbt=1.4.0
[2023-11-10T16:20:11.742-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:10  Unable to do partial parsing because saved manifest not found. Starting full parse.
[2023-11-10T16:20:11.742-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:10  Unable to do partial parsing because saved manifest not found. Starting full parse.
[2023-11-10T16:20:11.742-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Found 2 models, 0 tests, 0 snapshots, 0 analyses, 290 macros, 0 operations, 0 seed files, 1 source, 1 exposure, 0 metrics
[2023-11-10T16:20:11.742-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Found 2 models, 0 tests, 0 snapshots, 0 analyses, 290 macros, 0 operations, 0 seed files, 1 source, 1 exposure, 0 metrics
[2023-11-10T16:20:11.743-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.743-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.743-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Concurrency: 1 threads (target='dev')
[2023-11-10T16:20:11.743-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Concurrency: 1 threads (target='dev')
[2023-11-10T16:20:11.744-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.744-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.745-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  1 of 1 START sql table model main.movies_ratings_simplified .................... [RUN]
[2023-11-10T16:20:11.745-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  1 of 1 START sql table model main.movies_ratings_simplified .................... [RUN]
[2023-11-10T16:20:11.747-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  1 of 1 ERROR creating sql table model main.movies_ratings_simplified ........... [ERROR in 0.03s]
[2023-11-10T16:20:11.747-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  1 of 1 ERROR creating sql table model main.movies_ratings_simplified ........... [ERROR in 0.03s]
[2023-11-10T16:20:11.750-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.750-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.751-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Finished running 1 table model in 0 hours 0 minutes and 0.07 seconds (0.07s).
[2023-11-10T16:20:11.751-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Finished running 1 table model in 0 hours 0 minutes and 0.07 seconds (0.07s).
[2023-11-10T16:20:11.752-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.752-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.753-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Completed with 1 error and 0 warnings:
[2023-11-10T16:20:11.753-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Completed with 1 error and 0 warnings:
[2023-11-10T16:20:11.753-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.753-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.754-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Database Error in model movies_ratings_simplified (models/movies_ratings_simplified.sql)
[2023-11-10T16:20:11.754-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Database Error in model movies_ratings_simplified (models/movies_ratings_simplified.sql)
[2023-11-10T16:20:11.755-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11    no such table: imdb.movies_ratings
[2023-11-10T16:20:11.755-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11    no such table: imdb.movies_ratings
[2023-11-10T16:20:11.755-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11    compiled Code at target/run/simple/models/movies_ratings_simplified.sql
[2023-11-10T16:20:11.755-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11    compiled Code at target/run/simple/models/movies_ratings_simplified.sql
[2023-11-10T16:20:11.756-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.756-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11
[2023-11-10T16:20:11.756-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Done. PASS=0 WARN=0 ERROR=1 SKIP=0 TOTAL=1
[2023-11-10T16:20:11.756-0800] {subprocess.py:94} INFO - (astronomer-cosmos) - 00:20:11  Done. PASS=0 WARN=0 ERROR=1 SKIP=0 TOTAL=1
[2023-11-10T16:20:11.934-0800] {subprocess.py:98} INFO - (astronomer-cosmos) - Command exited with return code 1
[2023-11-10T16:20:11.934-0800] {subprocess.py:98} INFO - (astronomer-cosmos) - Command exited with return code 1
[2023-11-10T16:20:11.935-0800] {local.py:246} INFO - (astronomer-cosmos) - Inlets: []
[2023-11-10T16:20:11.935-0800] {local.py:246} INFO - (astronomer-cosmos) - Inlets: []
[2023-11-10T16:20:11.935-0800] {local.py:247} INFO - (astronomer-cosmos) - Outlets: []
[2023-11-10T16:20:11.935-0800] {local.py:247} INFO - (astronomer-cosmos) - Outlets: []
[2023-11-10T16:20:11.935-0800] {dag.py:2726} INFO - (astronomer-cosmos) - Sync 1 DAGs
[2023-11-10T16:20:11.948-0800] {dag.py:3486} INFO - (astronomer-cosmos) - Setting next_dagrun for example_cosmos_sources to 2023-11-10T00:00:00+00:00, run_after=2023-11-11T00:00:00+00:00
[2023-11-10T16:20:11.961-0800] {taskinstance.py:1847} ERROR - (astronomer-cosmos) - Task failed with exception
Traceback (most recent call last):
  File "/Users/justin.bandoro/astronomer-cosmos/cosmos/operators/local.py", line 455, in execute
    self.build_and_run_cmd(context=context, cmd_flags=cmd_flags)
  File "/Users/justin.bandoro/astronomer-cosmos/cosmos/operators/local.py", line 364, in build_and_run_cmd
    result = self.run_command(cmd=dbt_cmd, env=env, context=context)
  File "/Users/justin.bandoro/astronomer-cosmos/cosmos/operators/local.py", line 250, in run_command
    self.exception_handling(result)
  File "/Users/justin.bandoro/astronomer-cosmos/cosmos/operators/local.py", line 130, in exception_handling
    raise AirflowException(
airflow.exceptions.AirflowException: ('dbt command failed. The command returned a non-zero exit code 1. Details: ', '\x1b[0m00:20:10  Running with dbt=1.4.0', '\x1b[0m00:20:10  Unable to do partial parsing because saved manifest not found. Starting full parse.', '\x1b[0m00:20:11  Found 2 models, 0 tests, 0 snapshots, 0 analyses, 290 macros, 0 operations, 0 seed files, 1 source, 1 exposure, 0 metrics', '\x1b[0m00:20:11', "\x1b[0m00:20:11  Concurrency: 1 threads (target='dev')", '\x1b[0m00:20:11', '\x1b[0m00:20:11  1 of 1 START sql table model main.movies_ratings_simplified .................... [RUN]', '\x1b[0m00:20:11  1 of 1 ERROR creating sql table model main.movies_ratings_simplified ........... [\x1b[31mERROR\x1b[0m in 0.03s]', '\x1b[0m00:20:11', '\x1b[0m00:20:11  Finished running 1 table model in 0 hours 0 minutes and 0.07 seconds (0.07s).', '\x1b[0m00:20:11', '\x1b[0m00:20:11  \x1b[31mCompleted with 1 error and 0 warnings:\x1b[0m', '\x1b[0m00:20:11', '\x1b[0m00:20:11  \x1b[33mDatabase Error in model movies_ratings_simplified (models/movies_ratings_simplified.sql)\x1b[0m', '\x1b[0m00:20:11    no such table: imdb.movies_ratings', '\x1b[0m00:20:11    compiled Code at target/run/simple/models/movies_ratings_simplified.sql', '\x1b[0m00:20:11', '\x1b[0m00:20:11  Done. PASS=0 WARN=0 ERROR=1 SKIP=0 TOTAL=1')
```

**Comment by jbandoro** [comment]
@tatiana I was able to fix the sqlite integration test after some troubleshooting. The issue was that the current copying of directories was actually fixing a problem with the `simple` dbt project. Not using cosmos, running a model in the `simple` dbt project will fail with errors like:
```shell
❯ dbt run --models movies_ratings_simplified --profiles-dir /Users/justin.bandoro/astronomer-cosmos/dev/dags/dbt/simple --profile simple --target dev
20:16:24  Running with dbt=1.4.0
20:16:24  Unable to do partial parsing because profile has changed
20:16:24  Unable to do partial parsing because env vars used in profiles.yml have changed
20:16:24  Found 2 models, 0 tests, 0 snapshots, 0 analyses, 290 macros, 0 operations, 0 seed files, 1 source, 1 exposure, 0 metrics
20:16:24  
20:16:24  Concurrency: 1 threads (target='dev')
20:16:24  
20:16:24  1 of 1 START sql table model main.movies_ratings_simplified .................... [RUN]
20:16:24  1 of 1 ERROR creating sql table model main.movies_ratings_simplified ........... [ERROR in 0.02s]
20:16:24  
20:16:24  Finished running 1 table model in 0 hours 0 minutes and 0.06 seconds (0.06s).
20:16:24  
20:16:24  Completed with 1 error and 0 warnings:
20:16:24  
20:16:24  Database Error in model movies_ratings_simplified (models/movies_ratings_simplified.sql)
20:16:24    no such table: imdb.movies_ratings
20:16:24    compiled Code at target/run/simple/models/movies_ratings_simplified.sql
20:16:24  
20:16:24  Done. PASS=0 WARN=0 ERROR=1 SKIP=0 TOTAL=1
```
This is because there is a conflict in the `profiles.yml` setup for [sqlite](https://docs.getdbt.com/docs/core/connect-data-platform/sqlite-setup) where the `schema-directory` cannot have conflicts with the main schema. 

When cosmos is currently copying the entire dbt directory, what this did was have [2 different directories](https://github.com/astronomer/astronomer-cosmos/blob/6e41471712d9352524e70a59a1380e8e52c5ede2/dev/dags/dbt/simple/profiles.yml#L9C4-L9C5):
```yaml
      schema: 'main'
      schemas_and_paths:
        main: "{{ env_var('DBT_SQLITE_PATH', '.') }}/data/imdb.db"
      schema_directory: 'data'
``` 
The main schema which points to the original `imdb.db` location and the `schema-directory` which is the `data` directory in the copied directory, which is why running in cosmos works.

**Review by tatiana — APPROVED** [review]
Brilliant work, @jbandoro !

Thank you so much for digging into the SQLite issue and solving it. we'll be releasing this as part of 1.3. As soon as the tests pass, I'll merge this change and we'll have a new 1.3 alpha release.

**Comment by tatiana** [comment]
Thanks for the contribution, @jbandoro  , we released it as part of Cosmos 1.2.4 yesterday:
https://github.com/astronomer/astronomer-cosmos/releases/tag/astronomer-cosmos-v1.2.4

**Pull request merged**